File: catalog.cs
// catalog.cs — C# версия

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

class Book {
    public int Id { get; set; }
    public string ISBN { get; set; }
    public string Title { get; set; }
    public string Author { get; set; }
    public int Year { get; set; }
    public string Publisher { get; set; }
}

class BookCatalog {
    private List<Book> books = new List<Book>();
    private const string DataFile = "books.json";

    public BookCatalog() {
        Load();
    }

    private void Load() {
        if (File.Exists(DataFile)) {
            try {
                string json = File.ReadAllText(DataFile);
                books = JsonSerializer.Deserialize<List<Book>>(json) ?? new List<Book>();
            } catch {
                books = new List<Book>();
            }
        }
    }

    private void Save() {
        string json = JsonSerializer.Serialize(books, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(DataFile, json);
    }

    private bool IsValidISBN(string isbn) {
        isbn = isbn.Replace("-", "").Replace(" ", "");
        if (isbn.Length == 10) return CheckISBN10(isbn);
        if (isbn.Length == 13) return CheckISBN13(isbn);
        return false;
    }

    private bool CheckISBN10(string isbn) {
        if (!System.Text.RegularExpressions.Regex.IsMatch(isbn, @"^\d{9}[\dX]$")) return false;
        int sum = 0;
        for (int i = 0; i < 9; i++) {
            sum += (i + 1) * (isbn[i] - '0');
        }
        char check = isbn[9];
        if (check == 'X') sum += 100;
        else sum += 10 * (check - '0');
        return sum % 11 == 0;
    }

    private bool CheckISBN13(string isbn) {
        if (!System.Text.RegularExpressions.Regex.IsMatch(isbn, @"^\d{13}$")) return false;
        int sum = 0;
        for (int i = 0; i < 13; i++) {
            int digit = isbn[i] - '0';
            sum += (i % 2 == 0) ? digit : 3 * digit;
        }
        return sum % 10 == 0;
    }

    public int AddBook(string isbn, string title, string author, int year, string publisher) {
        if (!IsValidISBN(isbn)) throw new Exception("Неверный ISBN");
        if (books.Any(b => b.ISBN == isbn)) throw new Exception("Книга с таким ISBN уже существует");
        int id = books.Count + 1;
        books.Add(new Book { Id = id, ISBN = isbn, Title = title, Author = author, Year = year, Publisher = publisher });
        Save();
        return id;
    }

    public void ListAll() {
        if (books.Count == 0) {
            Console.WriteLine("\u001B[33mКаталог пуст.\u001B[0m");
            return;
        }
        Console.WriteLine($"\u001B[36m{"ID",-4} {"ISBN",-15} {"Название",-30} {"Автор",-20} {"Год",-6} {"Издательство",-20}\u001B[0m");
        Console.WriteLine(new string('-', 100));
        foreach (var b in books) {
            string title = b.Title.Length > 30 ? b.Title.Substring(0, 30) : b.Title;
            string author = b.Author.Length > 20 ? b.Author.Substring(0, 20) : b.Author;
            string pub = b.Publisher.Length > 20 ? b.Publisher.Substring(0, 20) : b.Publisher;
            Console.WriteLine($"{b.Id,-4} {b.ISBN,-15} {title,-30} {author,-20} {b.Year,-6} {pub,-20}");
        }
    }

    public Book SearchByISBN(string isbn) {
        isbn = isbn.Replace("-", "").Replace(" ", "");
        return books.FirstOrDefault(b => b.ISBN == isbn);
    }

    public List<Book> SearchByText(string text) {
        text = text.ToLower();
        return books.Where(b => b.Title.ToLower().Contains(text) || b.Author.ToLower().Contains(text)).ToList();
    }

    public bool Delete(int id) {
        var book = books.FirstOrDefault(b => b.Id == id);
        if (book != null) {
            books.Remove(book);
            Save();
            return true;
        }
        return false;
    }

    public void Edit(int id, string field, string value) {
        var book = books.FirstOrDefault(b => b.Id == id);
        if (book == null) throw new Exception("Книга не найдена");
        switch (field.ToLower()) {
            case "isbn":
                if (!IsValidISBN(value)) throw new Exception("Неверный ISBN");
                if (books.Any(b => b.Id != id && b.ISBN == value)) throw new Exception("ISBN уже используется");
                book.ISBN = value;
                break;
            case "title": book.Title = value; break;
            case "author": book.Author = value; break;
            case "year": book.Year = int.Parse(value); break;
            case "publisher": book.Publisher = value; break;
            default: throw new Exception("Неизвестное поле");
        }
        Save();
    }

    public void Stats() {
        if (books.Count == 0) {
            Console.WriteLine("Нет данных.");
            return;
        }
        var authors = books.GroupBy(b => b.Author).ToDictionary(g => g.Key, g => g.Count());
        var years = books.GroupBy(b => b.Year).ToDictionary(g => g.Key, g => g.Count());
        Console.WriteLine("\u001B[36m📊 Статистика:\u001B[0m");
        Console.WriteLine($"  Всего книг: {books.Count}");
        Console.WriteLine("  По авторам:");
        foreach (var kv in authors.OrderByDescending(kv => kv.Value)) {
            Console.WriteLine($"    {kv.Key}: {kv.Value}");
        }
        Console.WriteLine("  По годам:");
        foreach (var kv in years.OrderBy(kv => kv.Key)) {
            Console.WriteLine($"    {kv.Key}: {kv.Value}");
        }
    }

    public static void Main() {
        var catalog = new BookCatalog();
        while (true) {
            Console.WriteLine("\n\u001B[36m📚 Book Catalog (ISBN) (C#)\u001B[0m");
            Console.WriteLine("1. Добавить книгу");
            Console.WriteLine("2. Показать все книги");
            Console.WriteLine("3. Поиск по ISBN");
            Console.WriteLine("4. Поиск по названию/автору");
            Console.WriteLine("5. Удалить книгу");
            Console.WriteLine("6. Редактировать книгу");
            Console.WriteLine("7. Статистика");
            Console.WriteLine("8. Выход");
            Console.Write("Выберите действие: ");
            string choice = Console.ReadLine().Trim();
            switch (choice) {
                case "1":
                    Console.Write("ISBN (10 или 13 цифр): ");
                    string isbn = Console.ReadLine().Trim();
                    Console.Write("Название: ");
                    string title = Console.ReadLine().Trim();
                    Console.Write("Автор: ");
                    string author = Console.ReadLine().Trim();
                    Console.Write("Год: ");
                    int year = int.Parse(Console.ReadLine().Trim());
                    Console.Write("Издательство: ");
                    string pub = Console.ReadLine().Trim();
                    try {
                        int id = catalog.AddBook(isbn, title, author, year, pub);
                        Console.WriteLine($"\u001B[32m✅ Книга добавлена (ID: {id})\u001B[0m");
                    } catch (Exception e) {
                        Console.WriteLine($"\u001B[31m❌ Ошибка: {e.Message}\u001B[0m");
                    }
                    break;
                case "2": catalog.ListAll(); break;
                case "3":
                    Console.Write("Введите ISBN: ");
                    string searchIsbn = Console.ReadLine().Trim();
                    var book = catalog.SearchByISBN(searchIsbn);
                    if (book != null) {
                        Console.WriteLine($"ID: {book.Id}\nISBN: {book.ISBN}\nНазвание: {book.Title}\nАвтор: {book.Author}\nГод: {book.Year}\nИздательство: {book.Publisher}");
                    } else {
                        Console.WriteLine("\u001B[33mКнига не найдена.\u001B[0m");
                    }
                    break;
                case "4":
                    Console.Write("Введите название или автора: ");
                    string text = Console.ReadLine().Trim();
                    var results = catalog.SearchByText(text);
                    if (results.Any()) {
                        foreach (var b in results) {
                            Console.WriteLine($"{b.Id}: {b.Title} | {b.Author} | {b.Year}");
                        }
                    } else {
                        Console.WriteLine("\u001B[33mНичего не найдено.\u001B[0m");
                    }
                    break;
                case "5":
                    catalog.ListAll();
                    Console.Write("Введите ID для удаления: ");
                    int delId = int.Parse(Console.ReadLine().Trim());
                    if (catalog.Delete(delId)) {
                        Console.WriteLine("\u001B[32m✅ Книга удалена.\u001B[0m");
                    } else {
                        Console.WriteLine("\u001B[31m❌ Книга не найдена.\u001B[0m");
                    }
                    break;
                case "6":
                    catalog.ListAll();
                    Console.Write("Введите ID для редактирования: ");
                    int editId = int.Parse(Console.ReadLine().Trim());
                    Console.Write("Какое поле редактировать (isbn, title, author, year, publisher): ");
                    string field = Console.ReadLine().Trim().ToLower();
                    Console.Write("Новое значение: ");
                    string value = Console.ReadLine().Trim();
                    try {
                        catalog.Edit(editId, field, value);
                        Console.WriteLine("\u001B[32m✅ Книга обновлена.\u001B[0m");
                    } catch (Exception e) {
                        Console.WriteLine($"\u001B[31m❌ Ошибка: {e.Message}\u001B[0m");
                    }
                    break;
                case "7": catalog.Stats(); break;
                case "8": Console.WriteLine("До свидания!"); return;
                default: Console.WriteLine("\u001B[31mНеверный выбор.\u001B[0m"); break;
            }
        }
    }
}

File: catalog.java
// catalog.java — Java версия

import java.io.*;
import java.nio.file.*;
import java.util.*;

class Book {
    int id;
    String isbn;
    String title;
    String author;
    int year;
    String publisher;

    Book(int id, String isbn, String title, String author, int year, String publisher) {
        this.id = id;
        this.isbn = isbn;
        this.title = title;
        this.author = author;
        this.year = year;
        this.publisher = publisher;
    }

    String toJson() {
        return String.format("{\"id\":%d,\"isbn\":\"%s\",\"title\":\"%s\",\"author\":\"%s\",\"year\":%d,\"publisher\":\"%s\"}",
                id, isbn, title, author, year, publisher);
    }
}

public class catalog {
    private static List<Book> books = new ArrayList<>();
    private static final String DATA_FILE = "books.json";
    private static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        load();
        while (true) {
            System.out.println("\n\u001B[36m📚 Book Catalog (ISBN) (Java)\u001B[0m");
            System.out.println("1. Добавить книгу");
            System.out.println("2. Показать все книги");
            System.out.println("3. Поиск по ISBN");
            System.out.println("4. Поиск по названию/автору");
            System.out.println("5. Удалить книгу");
            System.out.println("6. Редактировать книгу");
            System.out.println("7. Статистика");
            System.out.println("8. Выход");
            System.out.print("Выберите действие: ");
            String choice = scanner.nextLine().trim();
            switch (choice) {
                case "1": addBook(); break;
                case "2": listAll(); break;
                case "3": searchByISBN(); break;
                case "4": searchByText(); break;
                case "5": deleteBook(); break;
                case "6": editBook(); break;
                case "7": stats(); break;
                case "8": System.out.println("До свидания!"); return;
                default: System.out.println("\u001B[31mНеверный выбор.\u001B[0m");
            }
        }
    }

    private static void load() {
        try {
            String content = new String(Files.readAllBytes(Paths.get(DATA_FILE)));
            // Упрощённо: если файл есть, парсим, иначе пустой список
            books = new ArrayList<>();
        } catch (IOException e) {
            books = new ArrayList<>();
        }
    }

    private static void save() {
        try {
            StringBuilder sb = new StringBuilder("[");
            for (int i = 0; i < books.size(); i++) {
                sb.append(books.get(i).toJson());
                if (i < books.size() - 1) sb.append(",");
            }
            sb.append("]");
            Files.write(Paths.get(DATA_FILE), sb.toString().getBytes());
        } catch (IOException e) {
            System.out.println("Ошибка сохранения.");
        }
    }

    private static boolean isValidISBN(String isbn) {
        isbn = isbn.replace("-", "").replace(" ", "");
        if (isbn.length() == 10) return checkISBN10(isbn);
        if (isbn.length() == 13) return checkISBN13(isbn);
        return false;
    }

    private static boolean checkISBN10(String isbn) {
        if (!isbn.matches("\\d{9}[\\dX]")) return false;
        int sum = 0;
        for (int i = 0; i < 9; i++) {
            sum += (i + 1) * (isbn.charAt(i) - '0');
        }
        char check = isbn.charAt(9);
        if (check == 'X') sum += 100;
        else sum += 10 * (check - '0');
        return sum % 11 == 0;
    }

    private static boolean checkISBN13(String isbn) {
        if (!isbn.matches("\\d{13}")) return false;
        int sum = 0;
        for (int i = 0; i < 13; i++) {
            int digit = isbn.charAt(i) - '0';
            sum += (i % 2 == 0) ? digit : 3 * digit;
        }
        return sum % 10 == 0;
    }

    private static void addBook() {
        System.out.print("ISBN (10 или 13 цифр): ");
        String isbn = scanner.nextLine().trim();
        System.out.print("Название: ");
        String title = scanner.nextLine().trim();
        System.out.print("Автор: ");
        String author = scanner.nextLine().trim();
        System.out.print("Год: ");
        int year = Integer.parseInt(scanner.nextLine().trim());
        System.out.print("Издательство: ");
        String publisher = scanner.nextLine().trim();

        if (!isValidISBN(isbn)) {
            System.out.println("\u001B[31m❌ Неверный ISBN.\u001B[0m");
            return;
        }
        for (Book b : books) {
            if (b.isbn.equals(isbn)) {
                System.out.println("\u001B[31m❌ Книга с таким ISBN уже существует.\u001B[0m");
                return;
            }
        }
        int id = books.size() + 1;
        books.add(new Book(id, isbn, title, author, year, publisher));
        save();
        System.out.println("\u001B[32m✅ Книга добавлена (ID: " + id + ")\u001B[0m");
    }

    private static void listAll() {
        if (books.isEmpty()) {
            System.out.println("\u001B[33mКаталог пуст.\u001B[0m");
            return;
        }
        System.out.printf("\u001B[36m%-4s %-15s %-30s %-20s %-6s %-20s\u001B[0m\n", "ID", "ISBN", "Название", "Автор", "Год", "Издательство");
        System.out.println("-".repeat(100));
        for (Book b : books) {
            String title = b.title.length() > 30 ? b.title.substring(0, 30) : b.title;
            String author = b.author.length() > 20 ? b.author.substring(0, 20) : b.author;
            String pub = b.publisher.length() > 20 ? b.publisher.substring(0, 20) : b.publisher;
            System.out.printf("%-4d %-15s %-30s %-20s %-6d %-20s\n", b.id, b.isbn, title, author, b.year, pub);
        }
    }

    private static void searchByISBN() {
        System.out.print("Введите ISBN: ");
        String isbn = scanner.nextLine().trim().replace("-", "").replace(" ", "");
        for (Book b : books) {
            if (b.isbn.equals(isbn)) {
                System.out.printf("ID: %d\nISBN: %s\nНазвание: %s\nАвтор: %s\nГод: %d\nИздательство: %s\n",
                        b.id, b.isbn, b.title, b.author, b.year, b.publisher);
                return;
            }
        }
        System.out.println("\u001B[33mКнига не найдена.\u001B[0m");
    }

    private static void searchByText() {
        System.out.print("Введите название или автора: ");
        String text = scanner.nextLine().trim().toLowerCase();
        List<Book> results = new ArrayList<>();
        for (Book b : books) {
            if (b.title.toLowerCase().contains(text) || b.author.toLowerCase().contains(text)) {
                results.add(b);
            }
        }
        if (results.isEmpty()) {
            System.out.println("\u001B[33mНичего не найдено.\u001B[0m");
            return;
        }
        for (Book b : results) {
            System.out.printf("%d: %s | %s | %d\n", b.id, b.title, b.author, b.year);
        }
    }

    private static void deleteBook() {
        listAll();
        System.out.print("Введите ID для удаления: ");
        int id = Integer.parseInt(scanner.nextLine().trim());
        Iterator<Book> it = books.iterator();
        while (it.hasNext()) {
            if (it.next().id == id) {
                it.remove();
                save();
                System.out.println("\u001B[32m✅ Книга удалена.\u001B[0m");
                return;
            }
        }
        System.out.println("\u001B[31m❌ Книга не найдена.\u001B[0m");
    }

    private static void editBook() {
        listAll();
        System.out.print("Введите ID для редактирования: ");
        int id = Integer.parseInt(scanner.nextLine().trim());
        Book target = null;
        for (Book b : books) {
            if (b.id == id) {
                target = b;
                break;
            }
        }
        if (target == null) {
            System.out.println("\u001B[31m❌ Книга не найдена.\u001B[0m");
            return;
        }
        System.out.print("Какое поле редактировать (isbn, title, author, year, publisher): ");
        String field = scanner.nextLine().trim().toLowerCase();
        System.out.print("Новое значение: ");
        String value = scanner.nextLine().trim();
        try {
            switch (field) {
                case "isbn":
                    if (!isValidISBN(value)) throw new Exception("Неверный ISBN");
                    for (Book other : books) {
                        if (other.id != id && other.isbn.equals(value)) throw new Exception("ISBN уже используется");
                    }
                    target.isbn = value;
                    break;
                case "title": target.title = value; break;
                case "author": target.author = value; break;
                case "year": target.year = Integer.parseInt(value); break;
                case "publisher": target.publisher = value; break;
                default: throw new Exception("Неизвестное поле");
            }
            save();
            System.out.println("\u001B[32m✅ Книга обновлена.\u001B[0m");
        } catch (Exception e) {
            System.out.println("\u001B[31m❌ Ошибка: " + e.getMessage() + "\u001B[0m");
        }
    }

    private static void stats() {
        if (books.isEmpty()) {
            System.out.println("Нет данных.");
            return;
        }
        Map<String, Integer> authors = new HashMap<>();
        Map<Integer, Integer> years = new HashMap<>();
        for (Book b : books) {
            authors.put(b.author, authors.getOrDefault(b.author, 0) + 1);
            years.put(b.year, years.getOrDefault(b.year, 0) + 1);
        }
        System.out.println("\u001B[36m📊 Статистика:\u001B[0m");
        System.out.printf("  Всего книг: %d\n", books.size());
        System.out.println("  По авторам:");
        for (Map.Entry<String, Integer> e : authors.entrySet()) {
            System.out.printf("    %s: %d\n", e.getKey(), e.getValue());
        }
        System.out.println("  По годам:");
        for (Map.Entry<Integer, Integer> e : years.entrySet()) {
            System.out.printf("    %d: %d\n", e.getKey(), e.getValue());
        }
    }
}
